great_tables/_options.py

```diff
@@ -1054,6 +1055,196 @@ def opt_table_outline(
+    the existing font stack. This way, if both fonts are not available, the table will fall back to
+    using the list of default table fonts. This behavior is controlled by the `add=` argument, which
+    is `True` by default.
+
+    With the `sza` dataset we'll create a two-column, eleven-row table. Within `opt_table_font()`,
+    the `stack=` argument will be supplied with the "rounded-sans" font stack. This sets up a family
+    of fonts with rounded, curved letterforms that should be locally available in different
+    computing environments.
+
+    ```{python}
+    from great_tables.data import sza
+
+    sza_mini = (
+        pl.from_pandas(sza)
+        .filter((pl.col("latitude") == "20") & (pl.col("month") == "jan"))
+        .drop_nulls()
+        .drop(["latitude", "month"])
+    )
+
+    (
+        GT(sza_mini)
+        .opt_table_font(stack="rounded-sans")
+        .opt_all_caps()
+    )
+    ```
+    """
+
+    if font is None and stack is None:
+        raise ValueError("Either `font=` or `stack=` must be provided.")
+
+    # Get the existing fonts for the table from the options; we may either prepend to this
+    # list or replace it entirely
+    existing_fonts = self._options.table_font_names.value
+
+    # If `existing_fonts` is not a list, throw an error
+    if not isinstance(existing_fonts, list):
+        raise ValueError("The value from `_options.table_font_names` must be a list.")
+
+    res = self
+
+    # If `stack` is provided, we ignore the `add` argument and always replace the existing fonts;
+    # This is because users are likely to want to use the font stack as a whole the set is safe
+    # enough to use across different systems (so keeping fallbacks usually won't be seen as
+    # necessary)
+    if stack is not None:
+        add = False
+
+    if font is not None:
+
+        # If `font` is a string, convert it to a list
+        if isinstance(font, str):
+            font = [font]
+
+    else:
+        font = []
+
+    if stack is not None:
+
+        # Case where value is given to `stack=` and this is a keyword that returns a
+        # list of fonts (i.e., the font stack); in this case we combine with `font=` values
+        # (if provided) and we *always* replace the existing fonts (`add=` is ignored)
+        from great_tables._helpers import system_fonts
+
+        font_stack = system_fonts(name=stack)
+
+        combined_fonts = font + font_stack
+        res = tab_options(res, table_font_names=combined_fonts)
+
+    else:
+
+        # Case where only `font=` is provided and that value is a list of fonts which may be
+        # prepended to the existing fonts or replaced entirely (depending on the `add=` value)
+        if add:
+            combined_fonts = font + existing_fonts
+            res = tab_options(res, table_font_names=combined_fonts)
+
+        else:
+            res = tab_options(res, table_font_names=font)
```

Comment:
It looks like..

* all 3 branches are about setting table_font_names
* 2 of the 3 branches are focused on creating a `combined_fonts` object

Could we try shifting the focus to creating `combined_fonts`, and then setting the table_font_names option after (vs focusing on setting the option)? Something like this...

```python
   if stack is not None:
    
        # Case where value is given to `stack=` and this is a keyword that returns a
        # list of fonts (i.e., the font stack); in this case we combine with `font=` values
        # (if provided) and we *always* replace the existing fonts (`add=` is ignored)
        from great_tables._helpers import system_fonts

        font_stack = system_fonts(name=stack)
        combined_fonts = font + font_stack
    elif add:
        # Case where `font=` is prepended to existing fonts
        combined_fonts = font + existing_fonts
    else:
        # Case where  `font=` replacing existing fonts
        combined_fonts = font
            
    res = tab_options(res, table_font_names=font)
```

Note two potentially useful pieces:

* the big if/elif/else supports the idea in the comment that if stack is set, then add is ignored.
* The comment in the previous else had two conditions ("depending on..."), and here is split into two (i.e. this is split up: `"may be prepended to the existing fonts or replaced entirely (depending on the add= value)"`)

Reply:
Thank you for the recommendation and the detailed explanations here (really helps!). I've made the change so that the conditionals are as above.